source/custom/visa_service.custom.cpp

```diff
@@ -447,32 +448,34 @@ ::grpc::Status VisaService::UsbControlIn(::grpc::ServerContext* context, const U
   if (context->IsCancelled()) {
     return ::grpc::Status::CANCELLED;
   }
-#if 1
-  return ::grpc::Status(grpc::StatusCode::DO_NOT_USE, "Custom code not implemented yet");
-#else
+  ViSession vi = VI_NULL;
+
   try {
     auto vi_grpc_session = request->vi();
-    ViSession vi = session_repository_->access_session(vi_grpc_session.name());
+    vi = session_repository_->access_session(vi_grpc_session.name());
     ViInt16 bm_request_type = (ViInt16)request->bm_request_type();
     ViInt16 b_request = (ViInt16)request->b_request();
     ViUInt16 w_value = request->w_value();
     ViUInt16 w_index = request->w_index();
     ViUInt16 w_length = request->w_length();
-    std::string buffer(return_count, '\0');
+    std::string* buffer = response->mutable_buffer();
+    buffer->resize(w_length);
     ViUInt16 return_count{};
-    auto status = library_->UsbControlIn(vi, bm_request_type, b_request, w_value, w_index, w_length, (ViByte*)buffer.data(), &return_count);
-    if (!status_ok(status)) {
+    auto status = library_->UsbControlIn(vi, bm_request_type, b_request, w_value, w_index, w_length, reinterpret_cast<ViByte*>(&buffer[0]), &return_count);
```

Comment:
Hang on, I don't think this is doing what you want. I think it needs to be `(&buffer->front())`

I think what it might be doing right now is treating buffer as an array of strings, taking the first string object, and casting that pointer to the desired type.

But if that's true, your calls to Read right now should be failing on the prebuild. The only thing that's clear to me right now is this kind of C++ code gets confusing quickly.

Reply:
And yes, that's exactly what happened. It crashed as soon as it called `WriteAndRead_Matches`.

Reply:
`&buffer->front()` still throws exception when it tries to read 0 byte, and extra logic will need to handle the 0 byte case. I will revert back to the C style type convert.

Reply:
Commit 4